authz/rbac_translator.go

```diff
@@ -0,0 +1,301 @@
+		Identifier: &v3rbacpb.Principal_OrIds{
+			OrIds: &v3rbacpb.Principal_Set{
+				Ids: principals,
+			},
+		},
+	}
+}
+
+func principalAnd(principals []*v3rbacpb.Principal) *v3rbacpb.Principal {
+	return &v3rbacpb.Principal{
+		Identifier: &v3rbacpb.Principal_AndIds{
+			AndIds: &v3rbacpb.Principal_Set{
+				Ids: principals,
+			},
+		},
+	}
+}
+
+func permissionOr(permission []*v3rbacpb.Permission) *v3rbacpb.Permission {
+	return &v3rbacpb.Permission{
+		Rule: &v3rbacpb.Permission_OrRules{
+			OrRules: &v3rbacpb.Permission_Set{
+				Rules: permission,
+			},
+		},
+	}
+}
+
+func permissionAnd(permission []*v3rbacpb.Permission) *v3rbacpb.Permission {
+	return &v3rbacpb.Permission{
+		Rule: &v3rbacpb.Permission_AndRules{
+			AndRules: &v3rbacpb.Permission_Set{
+				Rules: permission,
+			},
+		},
+	}
+}
+
+func getStringMatcher(value string) *v3matcherpb.StringMatcher {
+	if value == "*" {
+		return &v3matcherpb.StringMatcher{
+			MatchPattern: &v3matcherpb.StringMatcher_Prefix{
+				Prefix: "",
+			},
+		}
+	} else if strings.HasSuffix(value, "*") {
+		prefix := strings.TrimSuffix(value, "*")
+		return &v3matcherpb.StringMatcher{
+			MatchPattern: &v3matcherpb.StringMatcher_Prefix{
+				Prefix: prefix,
+			},
+		}
+	} else if strings.HasPrefix(value, "*") {
+		suffix := strings.TrimPrefix(value, "*")
+		return &v3matcherpb.StringMatcher{
+			MatchPattern: &v3matcherpb.StringMatcher_Suffix{
+				Suffix: suffix,
+			},
+		}
+	}
+	return &v3matcherpb.StringMatcher{
+		MatchPattern: &v3matcherpb.StringMatcher_Exact{
+			Exact: value,
+		},
+	}
+}
+
+func getHeaderMatcher(key, value string) *v3routepb.HeaderMatcher {
+	if value == "*" {
+		return &v3routepb.HeaderMatcher{
+			Name:                 key,
+			HeaderMatchSpecifier: &v3routepb.HeaderMatcher_PrefixMatch{PrefixMatch: ""},
+		}
+	} else if strings.HasSuffix(value, "*") {
+		prefix := strings.TrimSuffix(value, "*")
+		return &v3routepb.HeaderMatcher{
+			Name:                 key,
+			HeaderMatchSpecifier: &v3routepb.HeaderMatcher_PrefixMatch{PrefixMatch: prefix},
+		}
+	} else if strings.HasPrefix(value, "*") {
+		suffix := strings.TrimPrefix(value, "*")
+		return &v3routepb.HeaderMatcher{
+			Name:                 key,
+			HeaderMatchSpecifier: &v3routepb.HeaderMatcher_SuffixMatch{SuffixMatch: suffix},
+		}
+	}
+	return &v3routepb.HeaderMatcher{
+		Name:                 key,
+		HeaderMatchSpecifier: &v3routepb.HeaderMatcher_ExactMatch{ExactMatch: value},
+	}
+}
+
+func parsePrincipalNames(principalNames []string) []*v3rbacpb.Principal {
+	var or []*v3rbacpb.Principal
+	for _, principalName := range principalNames {
+		newPrincipalName := &v3rbacpb.Principal{
+			Identifier: &v3rbacpb.Principal_Authenticated_{
+				Authenticated: &v3rbacpb.Principal_Authenticated{
+					PrincipalName: getStringMatcher(principalName),
+				},
+			}}
+		or = append(or, newPrincipalName)
+	}
+	return or
+}
+
+func parsePeer(source peer) *v3rbacpb.Principal {
+	var and []*v3rbacpb.Principal
```

Comment:
I think you can refactor to eliminate this variable entirely.

Reply:
We could simplify this function to
```
if source.Principals != nil {
  principals := parsePrincipalNames(source.Principals)
  if len(principals) > 0 {
    return principalOr(principals)
  }
}
return &v3rbacpb.Principal{
  Identifier: &v3rbacpb.Principal_Any{
    Any:true,
  }  
}
```
The reason for writing it the way it is now, was to simplify the process of adding new fields and also to keep it consistent with C++ translator.
https://github.com/grpc/grpc/blob/master/src/core/lib/security/authorization/rbac_translator.cc

Let me know if you think we should avoid this complication, and simplify the function as shown above.

Reply:
> was to simplify the process of adding new fields

We shouldn't optimize for extension unless we're pretty certain such extension is coming.  As far as I'm aware, we weren't planning on ever changing this.  As such, clearer code should be more important than code that's more complex but easier to add new things into later.

Reply:
Sounds good. Simplified the method. Thank you!